sdk/core/core-client/src/interfaces.ts

```diff
@@ -724,6 +725,24 @@ export interface SpanConfig {
   namespace: string;
 }
 
+/**
+ * Used to configure additional policies added to the pipeline at construction.
+ */
+export interface AdditionalPolicyConfig {
+  /**
+   * A policy to be added.
+   */
+  policy: PipelinePolicy;
+  /**
+   * Determines if this policy be applied before or after retry logic.
+   * Only use `perRetry` if you need to modify the request again
+   * each time the operation is retried due to retryable service
+   * issues.
+   * Defaults to `perCall`
+   */
+  position?: "perCall" | "perRetry";
```

Comment:
This is really minor feedback, but this diverges from the other languages in a few small ways.

In .NET, `position` is a required parameter instead of optional, and the value is an enum rather than a string.  Should JS align more closely with this in either way, by making `position` a required parameter, or using an enum (or whatever is most similar to enums in JS) instead of a string?

```c#
public void AddPolicy(Azure.Core.Pipeline.HttpPipelinePolicy policy, Azure.Core.HttpPipelinePosition position) { }
```

In Java, `position` is specified on the policy itself, and I don't think JS would align with this.  However, the default position in Java is `PER_RETRY`, while JS chose the opposite `perCall`.  If JS does keep `position` as an optional parameter, I would lean towards defaulting to `perRetry`, in part to align with Java, and in part because I think `perRetry` is what more customers will want to use.

```java
default HttpPipelinePosition getPipelinePosition() {
    return HttpPipelinePosition.PER_RETRY;
}
```

A common scenario where you need to use `perRetry` is for services that add signatures to their HTTP requests, like Storage.  If a policy modifies a Storage request using `perCall`, it will appear to work until Storage needs to retry, where it will fail with an invalid signature.  This can be hard to catch during development since it only fails on a retry.

Reply:
> In .NET, `position` is a required parameter instead of optional, and the value is an enum rather than a string. Should JS align more closely with this in either way, by making `position` a required parameter, or using an enum (or whatever is most similar to enums in JS) instead of a string?

I'm not sure I see the value in making it a required parameter. For most user policies this shouldn't matter and it removes the convenience of being able to pass an array of policies directly.

We discourage TS enums in our API surface because they are not native to JS, nor are arbitrary integers meaningful when debugging. String literal unions are preferred whenever possible.

> In Java, `position` is specified on the policy itself, and I don't think JS would align with this. However, the default position in Java is `PER_RETRY`, while JS chose the opposite `perCall`. If JS does keep `position` as an optional parameter, I would lean towards defaulting to `perRetry`, in part to align with Java, and in part because I think `perRetry` is what more customers will want to use.

> A common scenario where you need to use `perRetry` is for services that add signatures to their HTTP requests, like Storage. If a policy modifies a Storage request using `perCall`, it will appear to work until Storage needs to retry, where it will fail with an invalid signature. This can be hard to catch during development since it only fails on a retry.

This isn't true. Today Storage always applies its signing policy **last**, even after the deserialization policy, so regardless of if you pick `perCall` or `perRetry` you will not be able to come after it with a user policy. Besides, if you came after it you *would* fail the signature as modifying the request post-signing will make it fail.

To my understanding, most `perRetry` logic would likely be policies that live as part of the client (such as signing) and we can ensure they are placed correctly. User policies usually only need to execute once per operation, since their modifications (e.g. disabling keepAlive) don't change on retry.

I don't feel that `perRetry` is necessarily a terrible default, I just don't see the justification and it again deviates from how our pipeline normally operates.

Reply:
Ack on string vs enum.

If `perRetry` policies are executed before signing, we have a different problem.  The perf framework needs to modify the request (by redirecting to the test-proxy) **after** signing, not before.  In all the other languages, I'm pretty sure `perRetry` runs after signing, since this is the only way our perf framework could work with Storage.

Before we merge this PR, I think we should consume this PR from the perf framework, and make sure it actually works to redirect requests from Storage to the test-proxy.

@HarshaNalluru: Can you work on this?

Reply:
I think your confusion is because storage is using `core-http` so your perf framework is using a custom `HttpClient` and not a policy as we had no way to do user policies back then.

Here's where the signing policy gets added for reference: https://github.com/Azure/azure-sdk-for-js/blob/6d4a3c7b9fc94c5a57b32fb9e1f6c59f30ffe5a9/sdk/storage/storage-blob/src/Pipeline.ts#L234

Reply:
Did some digging into .NET, the policy gets added as the last client retry policy here: https://github.com/Azure/azure-sdk-for-net/blob/4162f6fa2445b2127468b9cfd080f01c9da88eba/sdk/storage/Azure.Storage.Common/src/Shared/StorageClientOptions.cs#L114

And because customer retry policies get added after, it works the way you describe: https://github.com/Azure/azure-sdk-for-net/blob/4cd85e6dd23569269ccb6db92b813e0a394ee69d/sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs#L128

Reply:
Yes, as Jeff said, storage will be unaffected by this change.
Only the packages that use the core-v2 stack would pick this up.

Can test with a perf test project from such a package - example: form-recognizer.

Reply:
@HarshaNalluru: Assuming storage will upgrade to core-v2 at some point in the future, we need to make sure this can work with storage.  But this is really about all services, storage is just an example.  If form-recognizer uses the same authentication/signing pattern as storage, we might be able to use form-recognizer today as a test case.

@xirzec: Are you saying that JS currently works differently than .NET, where in JS customer retry policies would be added before SDK retry policies, but .NET is the opposite?  If so, this is another case where the languages really need to align.

Reply:
> @xirzec: Are you saying that JS currently works differently than .NET, where in JS customer retry policies would be added before SDK retry policies, but .NET is the opposite? If so, this is another case where the languages really need to align.

I mean yes, JS works differently in .NET in that we have a formal model for expressing ordering constraints and .NET has 2 client buckets and 2 customer buckets that get inserted in fixed positions, which can actually be enforced since clients can't bring their own pipeline or transport.

For the *default* pipeline, a bearer token auth policy gets added at the end of the standard phase if and only if a TokenCredential is passed in. For other auth methods, a client is welcome to manufacture their own pipeline and mix in their auth policies, or simply add them to the generated pipeline after the fact.

It sounds like what you're getting at though is you'd like a signing/auth phase so we could order policies after it? Or maybe this is the moral equivalent of .NET's "beforeTransport" ?

Reply:
The perf framework needs to redirect requests to test-proxy after authentication.  In .NET, Java, and Python this is done by adding a policy at position `perRetry`.  Best case, adding a policy at position `perRetry` would work the same in JS.  If this can't be done, then yes I think JS would need another position after authentication.

Reply:
I mean the main issue is that "authentication" isn't a formalized phase right now, nor do we force clients to use particular policies to accomplish it.

Reply:
@mikeharder I created #20129 to address the deficiency you identified. Assuming that merges first, I would update this one so that "perRetry" occurs `afterPhase: "Sign"` to align with .NET et al

Reply:
@mikeharder I pushed an update that accomplishes two things:

1. Makes position a required value
2. Ensures that perRetry happens post-signing.

I believe this puts us fully in-line with .NET, but let me know if you still have any concerns.

Reply:
@xirzec: Looks good to me.

@HarshaNalluru: Before we merge this PR, I think we should consume this PR from the perf framework, and make sure it actually works to redirect requests from Storage to the test-proxy.  Can you work on this?

Reply:
> @HarshaNalluru: Before we merge this PR, I think we should consume this PR from the perf framework, and make sure it actually works to redirect requests from Storage to the test-proxy. Can you work on this?

Are you planning to migrate Storage to corev2 as part of this experiment?

Reply:
I keep forgetting Storage is still on core-v1.  Is there another SDK on core-v2 which signs their requests we could use instead?  I would really like to verify this works end-to-end before merging.

Reply:
Yes, I wanted to get to this today.
Going after this now.

Reply:
Tested with tables, works fine. 😎

Adding a new method in the perf framework to be able to tweak the client options 
- https://github.com/Azure/azure-sdk-for-js/commit/725c2d988c01dff76d5aa28831698837d4b70fbf - new method and how it is used in tables

And here are the results without and with the test proxy enabled, redirection seems to happen just fine.

> npm run perf-test:node -- ListSimpleEntitiesTest
![image](https://user-images.githubusercontent.com/10452642/151877203-e3f63397-9c19-4cff-b0c4-8f19479cfe1b.png)

> npm run perf-test:node -- ListSimpleEntitiesTest --test-proxies http://localhost:5000
![image](https://user-images.githubusercontent.com/10452642/151877179-b5126084-143c-499f-a603-0d1e8515bb2b.png)

Reply:
Looks great, approving the PR.